### DISCUSSION TITLE
Hooks should be executed in order and in sequence in shake of determinism

### DISCUSSION THREAD

**Discussion by alfonso-presa** [webdriverio/webdriverio#6740] [Ideas]
**Environment (please complete the following information):**
 - **WebdriverIO version:**: v5 and v6
 - **Mode:** testrunner
 - **If WDIO Testrunner, running sync/async:** both
 - **Node.js version:** any
 - **NPM version:** any
 - **Browser name and version:** any
 - **Platform name and version:** any

**Config of WebdriverIO**
N/A

**Describe the bug**

Hooks are executed all in parallel leading for example to unexpected errors. Also the Promise.all, as it's fail fast, leads to the posiblity of hooks mixing ones with the others in case of a failure (for example if an afterTest fails, the other afters might keep executing.

Code for reference:
https://github.com/webdriverio/webdriverio/blob/1f639d2afdc626e26afcf5d23cb7237f446945f9/packages/wdio-utils/src/shim.js#L24

**Expected behavior**

Hooks should be executed in order and in sequence as declared in the services or the config. I propose changing the map for a reduce. Note it might have performance implications tough. I can hapilly PR if changing to reduce is ok for you.

**Comment by christian-bromann** [comment]
> Note it might have performance implications tough

Where do you see the performance implications?

**Comment by alfonso-presa** [comment]
Well.... before everything was being executed in parallel, so there might be some people performing tasks (data provisioning, mock service initialization, etc) that might take time. Now those tasks will be executed sequentially, with the corresponding time loss....

**Comment by mgrybyk** [comment]
I had also idea one day in the past to improve it somehow and even set order of execution: built-in hooks first, then service hooks, then user hooks. I also had issues because of that and because of all the hooks run in parallel

**Comment by christian-bromann** [comment]
> I also had issues because of that

Can you explain what kind of issues you had? I think no services should interfere with each other in any way which is way they have been executed in parallel in the first place.

**Comment by alfonso-presa** [comment]
En example of issue found because of this is performing `reloadSession` in a hook, the `sauce:context` executed by the sauce-service to set the scenario name is executed before the reloadSession.

I've also found issues trying to get the sessionId into the reporter because i had two different services for reloading and reporting... I ended up making only one service for both things, which IMHO lead to bad design....

**Comment by christian-bromann** [comment]
> the `sauce:context` executed by the sauce-service to set the scenario name is executed before the reloadSession.

That doesn't necessarily sound like a problem. There should be no race condition here. Hook nor `sauce:context` call run at the same time that could create a indeterminism behavior.

> i had two different services for reloading and reporting... I ended up making only one service for both things

I am not sure if a service should be used for reporting.

**Comment by alfonso-presa** [comment]
@christian-bromann 

> There should be no race condition here. Hook nor sauce:context call run at the same time that could create a indeterminism behavior.

The problem is that the `sauce:context` is executed against the old session and then after that session gets reloaded, so the new session lacks the category annotation. Check [first execution](https://app.eu-central-1.saucelabs.com/tests/36e0e6219f8b402aa47791056620af47) and [second execution](https://app.eu-central-1.saucelabs.com/tests/c1849b37bcbe4966b6253e423259fa33)

> I am not sure if a service should be used for reporting.

Sorry, I explained it wrongly. We use a hook to gather evidences in case or error (sources, screenshots, browser logs, etc) and add them to the allure report.

**Comment by christian-bromann** [comment]
> Check [first execution](https://app.eu-central-1.saucelabs.com/tests/36e0e6219f8b402aa47791056620af47) and [second execution](https://app.eu-central-1.saucelabs.com/tests/c1849b37bcbe4966b6253e423259fa33)

I think it depends where you do the reload. I will have to check what's the best way here.

> Sorry, I explained it wrongly. We use a hook to gather evidences in case or error (sources, screenshots, browser logs, etc) and add them to the allure report.

So there is one service to gather information and one that reports them to allure? I am not sure why it would be bad to have them in one service. One could argue that having services that need to be configured in a certain order is confusing and bad design. Note that you can run multiple things from different locations in one hook, e.g.:

```js
  // ...
  async afterTest () {
     await captureData()
     await reportData()
  }
  // ...
```

**Comment by alfonso-presa** [comment]
> I think it depends where you do the reload. I will have to check what's the best way here.

We are doing it in the `beforeScenario` with a flag to check if the scenario is the first one or not to avoid trash sessions. I was thinking of doing it in the after only if the scenario is not the last... but that's a lot more tricky to find out.

> So there is one service to gather information and one that reports them to allure?

There was one that reloads the sessions, and other that gathers evidences. The one gathering evidences adds the session id to the reports at the session start, and it was having race conditions with the one reloading, logging sometimes the wrong session id (the one from the previous test).

Of course this hooks can be redefined, and we could at the end find strategies to overcome this issue, but I think it would be useful to control the hook execution order to ease this things and make the system more predictable. The big problem here is that it might work almost always, so it leads to hard to find, possible source of flakiness, issues that I personally would not like to see in tests.
